pkg/verify/tlog.go

```diff
@@ -138,6 +141,78 @@ func VerifyTlogEntry(entity SignedEntity, trustedMaterial root.TrustedMaterial,
 	return verifiedTimestamps, nil
 }
 
+func VerifyArtifactTransparencyLogV2(entity SignedEntityWithRekorV2, trustedMaterial root.TrustedMaterial, logThreshold int) error { //nolint:revive
+	entries, err := entity.TlogV2Entries()
+	if err != nil {
+		return err
+	}
+	if len(entries) > maxAllowedTlogEntries {
+		return fmt.Errorf("too many tlog entries: %d > %d", len(entries), maxAllowedTlogEntries)
+	}
+	for i := 0; i < len(entries); i++ {
+		for j := i + 1; j < len(entries); j++ {
+			if entries[i].LogKeyID() == entries[j].LogKeyID() && entries[i].LogIndex() == entries[j].LogIndex() {
+				return fmt.Errorf("duplicate tlog entries found")
+			}
+		}
+	}
+
+	sigContent, err := entity.SignatureContent()
+	if err != nil {
+		return fmt.Errorf("getting signature content from signed entity: %w", err)
+	}
+	entitySignature := sigContent.Signature()
+
+	verificationContent, err := entity.VerificationContent()
+	if err != nil {
+		return fmt.Errorf("getting verification content from signed entity: %w", err)
+	}
+
+	logEntriesVerified := 0
+
+	for _, entry := range entries {
+		// TODO: validate hashedrekord/dsse obj?
+		rekorLogs := trustedMaterial.RekorLogs()
+		keyID := entry.LogKeyID()
+		hex64Key := hex.EncodeToString([]byte(keyID))
+		tlogVerifier, ok := rekorLogs[hex64Key]
+		if !ok {
+			return fmt.Errorf("could not find rekor log for public key")
+		}
+		sigVerifier, err := signature.LoadDefaultVerifier(tlogVerifier.PublicKey)
+		if err != nil {
+			return fmt.Errorf("loading verifier: %w", err)
+		}
+		noteVerifier, err := note.NewNoteVerifier("rekor-local", sigVerifier)
+		if err != nil {
+			return fmt.Errorf("loading note verifier: %w", err)
+		}
+		err = verify.VerifyLogEntry(entry.TransparencyLogEntry(), noteVerifier)
+		if err != nil {
+			return fmt.Errorf("verifying log entry: %w", err)
+		}
+		if !bytes.Equal(entry.Signature(), entitySignature) {
+			return fmt.Errorf("transparency log signature does not match")
+		}
+		protoPubKey := entry.PublicKey()
```

Comment:
IIRC this is confusingly named (to mirror rekor v1), this is a key _or_ certificate. It'd be nice to rename this or change this to return a verifier, and conditionally compare either a certificate or key to what's in the bundle.